skimage/transform/_thin_plate_splines.py

```diff
@@ -0,0 +1,320 @@
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4])
+    >>> np.allclose(yy_trans, expected_yy)
+    True
+    """
+
+    def __init__(self):
+        self._estimated = False
+        self.coefficients = None
+        self.src = None
+
+    def __call__(self, coords):
+        """Estimate the transformation from a set of corresponding points.
+
+        Parameters
+        ----------
+        coords : (N, 2) array_like
+            x, y coordinates to transform
+
+        Returns
+        -------
+        transformed_coords: (N, D) array
+            Destination coordinates
+        """
+        if self.coefficients is None:
+            raise ValueError(f"{self.coefficients}. Compute the `estimate`")
+        coeffs = self.coefficients
+        coords = np.array(coords)
+
+        if not coords.ndim == 2 or coords.shape[1] != 2:
+            raise ValueError("Input 'coords' must have shape (N,2)")
+
+        x_warp = self._spline_function(coords[:, 0], coords[:, 1], coeffs[:, 0])
+        y_warp = self._spline_function(coords[:, 0], coords[:, 1], coeffs[:, 1])
+        return np.vstack([x_warp, y_warp]).T
+
+    @property
+    def inverse(self):
+        raise NotImplementedError("This is yet to be implemented.")
+
+    def estimate(self, src, dst):
+        """Estimate optimal coefficients that describes the deformation of the points.
+
+
+        Parameters
+        ----------
+        src : (N, 2) array_like
+            Control point at source coordinates
+        dst : (N, 2) array_like
+            Control point at destination coordinates
+
+        Returns
+        -------
+        success: bool
+            True, indicates that the coefficients were successfully estimated.
+
+        Notes
+        -----
+        -  The number of source and destination points must match (N).
+        """
+
+        src = _ensure_2d(src)
+        dst = _ensure_2d(dst)
+
+        if src.shape != dst.shape:
+            raise ValueError("src and dst shape must be identical")
+
+        self.src = src
+        self.dst = dst
+        n, d = src.shape
+
+        K = self._radial_distance(src)
+        P = np.hstack([np.ones((n, 1)), src])
+        L = np.zeros((n + 3, n + 3), dtype=np.float32)
+        L[:n, :n] = K
+        L[:n, -3:] = P
+        L[-3:, :n] = P.T
+        V = np.concatenate([dst, np.zeros((d + 1, d))])
+        self.coefficients = np.dot(np.linalg.inv(L), V)
+        self._estimated = True
+        return self._estimated
+
+    def _spline_function(self, x, y, coeffs):
+        """Solve the spline function in the X and Y directions"""
+        w = coeffs[:-3]
+        a1, ax, ay = coeffs[-3:]
+        Pi_x = self.src[:, 0]
+        Pi_y = self.src[:, 1]
+
+        dx = Pi_x[:, np.newaxis] - x
+        dy = Pi_y[:, np.newaxis] - y
+        r = np.sqrt(dx**2 + dy**2)
+        radial_dist = _radial_basis_function(r)
```

Comment:
Above this function is called twice with 

```python
x_warp = self._spline_function(coords[:, 0], coords[:, 1], coeffs[:, 0])
y_warp = self._spline_function(coords[:, 0], coords[:, 1], coeffs[:, 1])
```

If I am not mistaken, the result of `radial_dist` is the same for both calls, so we could safe an entire call to `_radial_basis_function` if we calculate before and pass the result into `_spline_function`!

The part of this function that is actually different between both calls above is only 
```python
w = coeffs[:-3]
a1, ax, ay = coeffs[-3:]
# [...]
summation = np.dot(w, radial_dist)
return a1 + ax * x + ay * y + summation
```

Reply:
Yes, you are not mistaken. I created another function `_radial_distance` and pass the result into `_spline_function`.